src/sources/kafka.rs

```diff
@@ -251,20 +256,37 @@ async fn kafka_source(
     let mut stream = consumer.stream();
     let keys = Keys::from(log_schema(), &config);
 
+    let mut shutting_down = false;
     loop {
         tokio::select! {
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+            _ = &mut shutdown => {
+                if let Ok(topics) = consumer.subscription() {
+                    if let Err(error) = consumer.pause(&topics) {
+                        emit!(KafkaPauseResumeError { error });
                     }
                 }
+                shutting_down = true;
+
+                // If we have an ack stream, drop the sender end and allow acks to drain
+                match finalizer.take() {
+                    Some(finalizer) => drop(finalizer),
```

Comment:
Having merged master into this branch after https://github.com/vectordotdev/vector/pull/10434 landed, this no longer works! Since this reference is now an `Arc`, and there is a second reference to it now living in the consumer context, dropping it here doesn't close the sender channel anymore.

I don't see a way to get a mutable reference to the context to drop that second `finalizer` reference, but I think there are a couple other ways to proceed:

1. Refactor the `CustomContext` and the `post_rebalance` handler to use channels that notify up into this main loop, and call `finalizer.flush()` from here instead of passing a reference into the client
2. Add a `FinalizerSet.close` method that explicitly closes the sender channel

While 2) is probably a bit simpler to implement, I worry it could lead to some strange bugs where some source tries to add to a `FinalizerSet` that's been closed..I'll give 1) a try unless there are other suggestions?

Reply:
I'm realizing now that this entire issue might actually be a case of what @tobz described in the flush-acks-on-rebalance [PR comments](https://github.com/vectordotdev/vector/pull/10434/files#r785023759): acknowledgements left unprocessed before a shutdown or reload. Before, it could be caused by the shutdown signal ending the `ack_stream` immediately (the original issue I noticed), and now it can also happen around consumergroup rebalances. To restate here:

- We're processing N partitions and have in-flight messages from each of them
- (before, fixed by this PR: a shutdown signal ends the ack stream while there are pending acks to commit)
- The group rebalances and N/2 partitions are revoked. The source flushes all pending acknowledgements
- Vector reloads or restarts before consuming and acking new messages from the partition(s) it kept (the subsequent ack "covering" for the flushed away ones). So, if a restart/reload happens immediately after a rebalance event (perhaps during a deployment process that does a rolling restart/reload across all consumers), or if messages are infrequent and a restart happens any time before the next message.

I think the solution here is to track separate acknowledgement streams for each assigned partition, so that the source can flush acks only for revoked partitions. Something like tokio_stream's `StreamMap` is probably useful for this...it might be worth making a second PR for this fix though.